Cython/Utility/ImportExport.c

```diff
@@ -163,16 +171,42 @@ static PyObject *__Pyx_ImportDottedModuleRelFirst(PyObject *name, PyObject *part
 
 /////////////// Import.proto ///////////////
 
-static PyObject *__Pyx_Import(PyObject *name, PyObject *from_list, int level); /*proto*/
+static PyObject *__Pyx_Import(PyObject *name, PyObject *from_list, int level, PyObject *qualname); /*proto*/
 
 /////////////// Import ///////////////
-//@requires: ObjectHandling.c::PyObjectGetAttrStr
-//@requires:StringTools.c::IncludeStringH
+//@requires: StringTools.c::IncludeStringH
+//@requires: Builtins.c::HasAttr
+//@requires: ImportLookup
 
-static PyObject *__Pyx_Import(PyObject *name, PyObject *from_list, int level) {
+static PyObject *__Pyx_Import(PyObject *name, PyObject *from_list, int level, PyObject *qualname) {
     PyObject *module = 0;
     PyObject *empty_dict = 0;
     PyObject *empty_list = 0;
+    module = __Pyx__Import_LookupLock(qualname);
+    if (likely(module)) {
+        if (from_list) {
+            int module_ready = 1;
+            for (Py_ssize_t i = 0; i < __Pyx_PyList_GET_SIZE(from_list); i++) {
```

Comment:
```suggestion
            Py_ssize_t size = __Pyx_PyList_GET_SIZE(from_list);
            #if !CYTHON_ASSUME_SAFE_SIZE
            if (unlikely(size == -1)) {
                Py_DECREF(module);
                return NULL;
            }
            #endif
            for (Py_ssize_t i = 0; i < size; i++) {
```

in principle size can fail on limited api type builds and it isn't hugely quick there either (so probably worth taking out of the loop).